src/win32/posix_w32.c

```diff
@@ -42,12 +42,30 @@
 /* GetFinalPathNameByHandleW signature */
 typedef DWORD(WINAPI *PFGetFinalPathNameByHandleW)(HANDLE, LPWSTR, DWORD, DWORD);
 
-int p_ftruncate(int fd, long size)
+/**
+ * Truncate or extend file.
+ *
+ * We now take a "git_off_t" rather than "long" because
+ * files may be longer than 2Gb.
+ */
+int p_ftruncate(int fd, git_off_t size)
 {
+	if (size < 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
 #if defined(_MSC_VER) && _MSC_VER >= 1500
```

Comment:
I think that this (historic) test is incorrect.  `_chsize_s` was available way back in XP.  Should this just be `#ifdef __MINGW32__` ?